pageserver/src/page_cache.rs

```diff
@@ -226,51 +254,290 @@ pub struct CacheEntryContent {
     pub apply_pending: bool,
 }
 
+impl CacheEntryContent {
+    pub fn pack(&self, buf: &mut BytesMut) {
+        if let Some(image) = &self.page_image {
+            buf.put_u8(1);
+            buf.put_u16(image.len() as u16);
+            buf.put_slice(&image[..]);
+        } else if let Some(rec) = &self.wal_record {
+            buf.put_u8(0);
+            rec.pack(buf);
+        }
+    }
+    pub fn unpack(buf: &mut BytesMut) -> CacheEntryContent {
+        if buf.get_u8() == 1 {
+            let mut dst = vec![0u8; buf.get_u16() as usize];
+            buf.copy_to_slice(&mut dst);
+            CacheEntryContent {
+                page_image: Some(Bytes::from(dst)),
+                wal_record: None,
+                apply_pending: false,
+            }
+        } else {
+            CacheEntryContent {
+                page_image: None,
+                wal_record: Some(WALRecord::unpack(buf)),
+                apply_pending: false,
+            }
+        }
+    }
+}
+
 impl CacheEntry {
-    fn new(key: CacheKey) -> CacheEntry {
+    fn new(key: CacheKey, content: CacheEntryContent) -> CacheEntry {
         CacheEntry {
             key,
-            content: Mutex::new(CacheEntryContent {
-                page_image: None,
-                wal_record: None,
-                apply_pending: false,
-            }),
+            content: Mutex::new(content),
             walredo_condvar: Condvar::new(),
         }
     }
 }
 
-#[derive(Eq, PartialEq, Hash, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Hash, Ord, Clone, Copy)]
 pub struct RelTag {
     pub spcnode: u32,
     pub dbnode: u32,
     pub relnode: u32,
     pub forknum: u8,
 }
 
-#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
+impl RelTag {
+    pub fn pack(&self, buf: &mut BytesMut) {
+        buf.put_u32(self.spcnode);
+        buf.put_u32(self.dbnode);
+        buf.put_u32(self.relnode);
+        buf.put_u32(self.forknum as u32);
+    }
+    pub fn unpack(buf: &mut BytesMut) -> RelTag {
+        RelTag {
+            spcnode: buf.get_u32(),
+            dbnode: buf.get_u32(),
+            relnode: buf.get_u32(),
+            forknum: buf.get_u32() as u8,
```

Comment:
When I see `as` used to truncate integers I usually ask whether there's a better way:
- use the smaller type in the first place
- use `try_into()` if we expect it to fit without truncation
If truncation is really what is needed, can you add a comment explaining why?